57 add social auth

Abdurahman-hassan opened the pull request:
### Description:
This pull request introduces a custom-built, multi-platform OAuth integration system that currently supports Google and Facebook. This integration is designed without relying on external OAuth libraries, ensuring full control over the authentication process and easier compliance with specific security policies.

Details:

- BaseOAuthHandler Class: Central to our approach, this class in base_oauth_handler.py serves as the blueprint for handling OAuth interactions across any platform. It manages token extraction, user information retrieval, and user creation.

- Provider-Specific Handlers: GoogleHandler and FacebookHandler, extending BaseOAuthHandler, are tailored to meet the specific API requirements of Google and Facebook respectively, including unique user info endpoints and response handling.

- OAuth Factory Function: Located in oauth_factory.py, this function dynamically selects the appropriate handler based on the provider specified, enabling the system to adapt to various OAuth providers without hard-coding dependencies.
OAuthLoginView: This view leverages the OAuth factory to process authentication requests, providing a unified interface that can handle any supported provider transparently.

- The system’s architecture allows for the addition of more providers (like Twitter, LinkedIn, etc.) by simply extending BaseOAuthHandler and registering the new handler in the factory function.

- No external OAuth libraries are used, which avoids additional dependencies and simplifies compliance with data handling regulations. To create a scalable and easily extendable system that can incorporate additional OAuth providers as needed without extensive reconfiguration or additional dependencies.

### Related Issue
#57 

### Screenshots (if applicable)
N/A

Abdurahman-hassan committed:
add label and verbose_name for app.account in order to avoid conflict with the allauth app

Signed-off-by: abdelrahman hassan <abdelrahman_hassan31@yahoo.com>

Abdurahman-hassan committed:
Add ENABLE_SOCIAL_LOGIN in order to prevent the social urls as per the configurations inside env file, fix linting

Signed-off-by: abdelrahman hassan <abdelrahman_hassan31@yahoo.com>

Abdurahman-hassan committed:
Add auth/social in urls

Signed-off-by: abdelrahman hassan <abdelrahman_hassan31@yahoo.com>

Abdurahman-hassan committed:
add BaseOAuthHandler class

Signed-off-by: abdelrahman hassan <abdelrahman_hassan31@yahoo.com>

Abdurahman-hassan committed:
add oauth_factory file

Signed-off-by: abdelrahman hassan <abdelrahman_hassan31@yahoo.com>

Abdurahman-hassan committed:
add providers_handler file

Signed-off-by: abdelrahman hassan <abdelrahman_hassan31@yahoo.com>

Abdurahman-hassan committed:
add OAuthLogin generic view, inside this view we can handle multiple providers

Signed-off-by: abdelrahman hassan <abdelrahman_hassan31@yahoo.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wissam-al-kahwaji commented on config/settings/base.py:
```diff
@@ -65,10 +64,9 @@
 LOCAL_APPS = [
     "app.user",
-    "app.account",
+    "app.account.apps.AccountConfig",
```
remove apps.AccountConfig,
use Just "app.account"

wissam-al-kahwaji commented on app/account/apps.py:
```diff
@@ -4,3 +4,5 @@
 class AccountConfig(AppConfig):
     default_auto_field = "django.db.models.BigAutoField"
     name = "app.account"
+    label = "app_account"  # in order to avoid conflict with the allauth app
```
line [7,8], This also causes errors and should be removed.

Abdurahman-hassan committed:
Fix: OWASP: Log Injection.
     Common Weakness Enumeration: CWE-117,

     by adding sanitize_message function inside our custome Logger class,

     This function Sanitize the message by removing newlines, tabs,
                               and non-alphanumeric characters, and replacing them with spaces.
using our custome logger class insted of the logger base class

Signed-off-by: abdelrahman hassan <abdelrahman_hassan31@yahoo.com>

Abdurahman-hassan committed:
Merge branch 'refs/heads/develop' into 57-add-social-auth

# Conflicts:
#	config/utils/logging/general.py

Abdurahman-hassan committed:
return logger back

Signed-off-by: abdelrahman hassan <abdelrahman_hassan31@yahoo.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Abdurahman-hassan committed:
renaming app insted of APP

Signed-off-by: abdelrahman hassan <abdelrahman_hassan31@yahoo.com>

Abdurahman-hassan committed:
remove testing logs

Signed-off-by: abdelrahman hassan <abdelrahman_hassan31@yahoo.com>

Abdurahman-hassan committed:
removing non-alphanumeric characters, include sanitize_message directly inside logs fun

Signed-off-by: abdelrahman hassan <abdelrahman_hassan31@yahoo.com>

Abdurahman-hassan committed:
Merge remote-tracking branch 'origin/57-add-social-auth' into 57-add-social-auth

# Conflicts:
#	app/account/views/social.py
#	config/utils/logging/general.py

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Abdurahman-hassan commented:
conflict, pytest, codeQL has been fixed.
it require to use built it replace function to pass here, but i used enhanced regex inside our custom logger class
it's a fake error

wissam-al-kahwaji commented on config/utils/logging/general.py:
```diff
@@ -0,0 +1,21 @@
+from config.settings.base import app
+
+
+class Logger:
```
Why did you create a logger here?
He's already inside 
octo.logging

https://github.com/oc-to/octo-framework

wissam-al-kahwaji commented on app/utils/providers_handler.py:
```diff
@@ -0,0 +1,18 @@
+        super().__init__(access_token)
+        self.USER_INFO_URL = self.get_user_info_url()
+
+    def get_user_info_url(self):
```
It is not correct to include this here since we are treating **BaseOAuthHandler** as a pattern so it must already exist, and can be used or ignored.

wissam-al-kahwaji commented:
Given the time, I will complete this task myself. Thank you for your valuable contribution. It was a very good contribution

wissam-al-kahwaji reviewed (approved):
Given the time, I will complete this task myself. Thank you for your valuable contribution. It was a very good contribution
